File: binary_tree_from_inorder_and_postorder_traversal.py
"""
Problem: 106. Construct Binary Tree from Inorder and Postorder Traversal
Leetcode: https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
Solution: Recursion
Time Complexity: O(n), we traverse all nodes in the tree
Space Complexity: O(n), since we store the entire tree.
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # class variables
    inorder = []
    postorder = []
    index_map = {}
    post_idx = len(postorder) - 1

    def buildTree(self, inorder, postorder) -> TreeNode:
        self.inorder = inorder
        self.postorder = postorder

        # Create a hashmap with inorder values and indices
        idx = 0
        for i in range(len(inorder)):
            self.index_map[inorder[i]] = i
        return self.helper(0, len(inorder) - 1)

    def helper(self, left, right):
        # Base case
        if left > right:
            return None

        # find the root from postorder list
        root_val = self.postorder[self.post_idx]
        index = self.index_map[root_val]
        self.post_idx -= 1

        root = TreeNode(root_val)

        # Because this is postorder traversal, we've to build right side first
        root.right = self.helper(index + 1, right)
        root.left = self.helper(left, index - 1)
        return root

File: BinaryTreeFromInOrderAndPostOrderTraversal.java
//TC: O(n)
//SC:O(n)
//works in leetcode
//last element in the postorder array is the root
//use decrementer to make nodes
//use inorder to determine the relationship of node. Left subtree and right subtree is created and assigned.
public class BinaryTreeFromInOrderAndPostOrderTraversal {
    private static int postorderIndex;
    private static int[] postorder,  inorder;
    private static Map<Integer, Integer> postorderMap;


    public static  TreeNode buildTree(int[] postorder1, int[] inorder1) {
        postorderIndex=postorder1.length-1;
        postorder= postorder1;
        inorder=inorder1;
        postorderMap = new HashMap<>();
        for(int i=0;i< inorder.length;i++){
            postorderMap.put(inorder[i], i);
        }
        return buildTree(0, inorder.length-1);
    }

    private static TreeNode buildTree(int inOrderStart, int inOrderEnd){

        //base
        if(inOrderStart > inOrderEnd)
            return null;

        //recursion
        TreeNode node = new TreeNode(postorder[postorderIndex]);
        postorderIndex--;

        int inorderIndex = getIndex(node.val);
        node.right = buildTree(inorderIndex+1, inOrderEnd);
        node.left = buildTree(inOrderStart, inorderIndex-1);

        return node;

    }

    private static int getIndex(int element){
        return postorderMap.get(element);
    }



    public static void main(String [] args){
        int [] postorder = {9,15,7,20,3};
        int [] inorder = {9,3,15,20,7};
        TreeNode tree =  buildTree(postorder, inorder);
        System.out.println(tree.val);
    }
}
